pilot/pkg/proxy/envoy/v2/ads_common.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config/schemas"
+)
+
+func ProxyNeedsPush(proxy *model.Proxy, pushEv *XdsEvent) bool {
+	if !features.ScopePushes.Get() {
+		// If push scoping is not enabled, we push for all proxies
+		return true
+	}
+
+	targetNamespaces := pushEv.namespacesUpdated
+	configs := pushEv.configTypesUpdated
+
+	// appliesToProxy starts as false, we will set it to true if we encounter any configs that require a push
+	appliesToProxy := false
+	// If no config specified, this request applies to all proxies
+	if len(configs) == 0 {
+		appliesToProxy = true
+	}
+Loop:
+	for config := range configs {
+		switch config {
+		case schemas.Gateway.Type:
+			if proxy.Type == model.Router {
+				return true
+			}
+		case schemas.QuotaSpec.Type, schemas.QuotaSpecBinding.Type:
+			if proxy.Type == model.SidecarProxy {
+				return true
+			}
+		default:
+			appliesToProxy = true
+			break Loop
+		}
+	}
+
+	if !appliesToProxy {
+		return false
+	}
+
+	// If no only namespaces specified, this request applies to all proxies
+	if len(targetNamespaces) == 0 {
+		return true
+	}
+
+	// Otherwise, only apply if the egress listener will import the config present in the update
+	for ns := range targetNamespaces {
+		if proxy.SidecarScope.DependsOnNamespace(ns) {
+			return true
+		}
+	}
+	return false
+}
+
+type XdsType int
+
+const (
+	CDS XdsType = iota
+	EDS
+	LDS
+	RDS
+)
+
+// TODO: merge with ProxyNeedsPush
+func PushTypeFor(proxy *model.Proxy, pushEv *XdsEvent) map[XdsType]bool {
+	out := map[XdsType]bool{}
+
+	// In case configTypes is not set, for example mesh configuration updated.
+	// If push scoping is not enabled, we push all xds
+	if !features.ScopePushes.Get() || len(pushEv.configTypesUpdated) == 0 {
+		out[CDS] = true
+		out[EDS] = true
+		out[LDS] = true
+		out[RDS] = true
+		return out
+	}
+
+	if proxy.Type == model.SidecarProxy {
+		for config := range pushEv.configTypesUpdated {
+			switch config {
+			case schemas.VirtualService.Type:
+				out[LDS] = true
+				out[RDS] = true
+			case schemas.Gateway.Type:
+				// Do not push
+			case schemas.ServiceEntry.Type, schemas.SyntheticServiceEntry.Type:
```

Comment:
is there a way to figure out instance update for a service entry? In which case we can only send EDS when instance changes?

Reply:
When SyntheticServiceEntry controller enabled, it can do that. But not work for the `ServiceEntry` i think.
 For eds only, this function is not called. It goes through https://github.com/istio/istio/blob/920212422057448118b49f743c8a568d2138172a/pilot/pkg/proxy/envoy/v2/ads.go#L494-L506

Reply:
Yeah, that is because the instance handler here https://github.com/istio/istio/blob/920212422057448118b49f743c8a568d2138172a/pilot/pkg/proxy/envoy/v2/discovery.go#L170 does not set the `edsUpdatedServices` and it goes through full push

Reply:
We have no way to distinguish from a simple SE update event whether it is endpoints change or service change based on the code now. But seems not so complex to do that.

Reply:
Yeah, not complicated, but we need to rework the handlers - currently event handler is blindly called whether it is add/update https://github.com/istio/istio/blob/master/pilot/pkg/config/kube/crd/controller/controller.go#L191 and we have to pass additional data to event handler